experimental/dds/tree2/src/test/tree/anchorSet.spec.ts

```diff
@@ -349,6 +350,104 @@ describe("AnchorSet", () => {
+		const insertMark: Delta.Insert = {
+			type: Delta.MarkType.Insert,
+			content: [node].map(singleTextCursor),
+		};
+		const deleteMark: Delta.Remove = {
+			type: Delta.MarkType.Remove,
+			count: 1,
+			detachId,
+		};
+		const replaceMark: Delta.Insert = {
+			type: Delta.MarkType.Insert,
+			content: [node].map(singleTextCursor),
+			oldContent: { detachId: { minor: 42 } },
+		};
+		const anchors = new AnchorSet();
+
+		// Insert a node at the root to set up listeners on it
+		announceTestDelta(
+			makeDelta(insertMark, makePath([rootFieldKey, 0], [fieldFoo, 3])),
+			anchors,
+		);
+		const anchor0 = anchors.track(makePath([rootFieldKey, 0]));
+		const node0 = anchors.locate(anchor0) ?? assert.fail();
+
+		let beforeCounter = 0;
+		let afterCounter = 0;
+
+		const unsubscribeBeforeChange = node0.on("beforeChange", (n: AnchorNode) => {
+			beforeCounter++;
+			assert.strictEqual(afterCounter, beforeCounter - 1, "beforeChange fired out of order");
+		});
+		const unsubscribeAfterChange = node0.on("afterChange", (n: AnchorNode) => {
+			afterCounter++;
+			assert.strictEqual(afterCounter, beforeCounter, "afterChange fired out of order");
+		});
+
+		// Test an insert delta
+		announceTestDelta(
+			makeDelta(insertMark, makePath([rootFieldKey, 0], [fieldFoo, 4])),
+			anchors,
+		);
+		assert.strictEqual(beforeCounter, 1);
+		assert.strictEqual(afterCounter, 1);
+
+		// Test a replace delta
+		announceTestDelta(
+			makeDelta(replaceMark, makePath([rootFieldKey, 0], [fieldFoo, 5])),
+			anchors,
+		);
+		assert.strictEqual(beforeCounter, 2);
+		assert.strictEqual(afterCounter, 2);
+
+		// Test a delete delta
+		announceTestDelta(
+			makeDelta(deleteMark, makePath([rootFieldKey, 0], [fieldFoo, 5])),
+			anchors,
+		);
+		assert.strictEqual(beforeCounter, 3);
+		assert.strictEqual(afterCounter, 3);
+
+		// Test a move delta
+		const moveOutMark: Delta.MoveOut = {
+			type: Delta.MarkType.MoveOut,
+			count: 1,
+			moveId: brand(1),
+		};
+		const moveInMark: Delta.MoveIn = {
+			type: Delta.MarkType.MoveIn,
+			count: 1,
+			moveId: moveOutMark.moveId,
+		};
+		const moveDelta = makeFieldDelta(
+			[moveOutMark, 1, moveInMark],
+			makeFieldPath(fieldFoo, [rootFieldKey, 0]),
+		);
+		announceTestDelta(moveDelta, anchors);
+		// TODO: this is wrong. Currently a move triggers the events twice, once when detaching nodes from the source
+		// and again when attaching them at the destination. Couldn't think of a quick way to avoid that.
+		// This means that:
+		// - In the first beforeChange callback, the moved nodes are still at the source (correct).
+		// - In the first afterChange callback, the moved nodes are not in the tree (probably unexpected).
+		// - In the second beforeChange callback, the moved nodes are not in the tree (probably unexpected).
+		// - In the second afterChange callback, the moved nodes are at the destination (correct).
+		assert.strictEqual(beforeCounter, 5);
+		assert.strictEqual(afterCounter, 5);
```

Comment:
Particularly interested if anyone has thoughts on this. I'm not sure that the current information we pass to the visitor is enough to make this behavior work in a way that I think would be the most natural to users of editable tree, i.e., if nodes are moving you get one beforeChange trigger when the nodes are still at their original location, and one afterChange trigger when they're at their destination. Otherwise it feels like we overexpose the internal machinery.

Reply:
While it may indeed be a little suprising, this is a limitation we planned on embracing so this is working as intended.

More context: for every atomic editing operation supported by ST, we have to decide whether we commit to notifying listeners in a way that is also atomic. We haven't committed to doing that for moves because moves affect multiple (two) locations in the tree, and we communicating such an operation atomically would introduce a level of complexity we think is not worth it.

It would good to call this out in the doc comments for beforeChange/afterChange events.

Reply:
Ok, as long as we're deliberately making that choice. I'll document the events themselves and will try to leave a mental note in everyone's minds that this probably wants to be called out in more public-facing documentation as well.